### PR TITLE
mpi ops& distributed batchnorm

### PR DESCRIPTION
**Pull request by Gword**

### CODE REVIEW

File: extern/mpi/ops/mpi_reduce_op.cc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+#else // JIT
+#ifdef JIT_cpu
+void MpiReduceOp::jit_run() {
+    @for(i, 0, XDIM, index_t xshape@i = x->shape[@i];)
+    int size = 1 @for(i, 0, XDIM,  * xshape@{i});
+    auto* __restrict__ xp = x->ptr<Tx>();
+    auto* __restrict__ yp = y->ptr<Tx>();
+    index_t num = y->num;
+    for (index_t i=0; i<num; i++)
```
Comment (cjld) [inline extern/mpi/ops/mpi_reduce_op.cc:62]: if (root != mpi_world_rank) for ....

File: extern/mpi/ops/mpi_broadcast_op.cc [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    @for(i, 0, XDIM, index_t xshape@i = x->shape[@i];)
+    int size = 1 @for(i, 0, XDIM,  * xshape@{i});
+    auto* __restrict__ xp = x->ptr<Tx>();
+    auto* __restrict__ yp = y->ptr<Tx>();
+    if (mpi_world_rank == root) {
+        for (int i = 0; i < mpi_world_size; i++) {
+            MPI_Send(xp, size, MPI_FLOAT, i, 0, MPI_COMM_WORLD);
+        }
+    }
+    MPI_Recv(yp, size, MPI_FLOAT, root, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
```
Comment (cjld) [inline extern/mpi/ops/mpi_broadcast_op.cc:66]: MPI_broadcast?

File: extern/mpi/ops/mpi_broadcast_op.cc [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            return;
+        }
+    }
+    #endif
+    y = create_output(nullptr, x->dtype());
+    ASSERT(x->dtype().is_float());
+}
+
+void MpiBroadcastOp::infer_shape() {
+    y->set_shape(x->shape);
```
Comment (cjld) [inline extern/mpi/ops/mpi_broadcast_op.cc:40]:
```
if (root == mpi_world_rank)
  y->share_with(x);
```

File: extern/mpi/ops/mpi_reduce_op.cc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        yp[i] = 0;
+    MPI_Reduce(xp, yp, size, MPI_FLOAT, MPI_SUM, root, MPI_COMM_WORLD);
```
Comment (cjld) [inline extern/mpi/ops/mpi_reduce_op.cc:64]: 支持多种类型，
参考https://github.com/Jittor/jittor/pull/57/files#diff-52bb7f0a5beaae1676a30579a7361f3bR46

File: python/jittor/nn.py [hunk 5]
```diff
@@ -255,9 +270,14 @@ def __init__(self, num_features, eps=1e-5, momentum=0.1, affine=None, is_train=T
         self.running_var = init.constant((num_features,), "float32", 1.0).stop_grad()
 
     def execute(self, x):
+        mpi = jt.compile_extern.mpi
         if self.is_train:
             xmean = jt.mean(x, dims=[0,2,3], keepdims=1)
             x2mean = jt.mean(x*x, dims=[0,2,3], keepdims=1)
+            if self.sync and jt.compile_extern.mpi_ops is not None:
+                xmean =  jt.compile_extern.mpi_ops.mpi_all_reduce(xmean)/jt.compile_extern.mpi.world_size()
+                x2mean =  jt.compile_extern.mpi_ops.mpi_all_reduce(x2mean)/jt.compile_extern.mpi.world_size()
```
Comment (cxjyxxme) [inline python/jittor/nn.py:279]: 不同进程的batch数量可能不一样，好像不能直接除

File: python/jittor/test/test_mpi_batchnorm.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        return norm_x * w + b
+
+@unittest.skipIf(mpi is None, "no inside mpirun")
+class TestMpiBatchnorm(unittest.TestCase):
+    @classmethod
+    def setUpClass(self):
+        np.random.seed(0)
+        jt.seed(3)
+
+    def test_batchnorm(self):
```
Comment (cxjyxxme) [inline python/jittor/test/test_mpi_batchnorm.py:56]: 加个测试下不同进程batchsize不一样的case

**Review by cxjyxxme — CHANGES_REQUESTED** [review]

File: extern/mpi/ops/mpi_all_reduce_op.cc [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+#ifdef JIT_cpu
+void MpiAllReduceOp::jit_run() {
+    @define(T_MPI,
+        @if(@strcmp(@Tx,float)==0 || @strcmp(@Tx,float32)==0, MPI_FLOAT)
+        @if(@strcmp(@Tx,int)==0 || @strcmp(@Tx,int32)==0, MPI_INT)
+        @if(@strcmp(@Tx,float64)==0 || @strcmp(@Tx,double)==0, MPI_DOUBLE)
+        @if(@strcmp(@Tx,int64)==0, MPI_DOUBLE_INT)
+    )
+    @define(OP_MPI,
+        @if(@strcmp(@OP,add)==0, MPI_SUM)
```
[the remark this answers is not in the document]
Comment (cjld) [reply extern/mpi/ops/mpi_all_reduce_op.cc:76]: MPI_MIN,MPI_MAX may added in the future

File: extern/mpi/ops/mpi_all_reduce_op.cc [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            : nullptr;
+        if (nccl_all_reduce) {
+            auto var = nccl_all_reduce(x);
+            forward(var);
+            return;
+        }
+    }
+    #endif
+    y = create_output(nullptr, x->dtype());
+    ASSERT(x->dtype().is_float());
```
[the remark this answers is not in the document]
Comment (cjld) [reply extern/mpi/ops/mpi_all_reduce_op.cc:48]: forget to remove, removed now

File: extern/mpi/ops/mpi_all_reduce_op.cc [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+VarPtr MpiAllReduceOp::grad(Var* out, Var* dout, Var* v, int v_index) {
+    static auto mpi_all_reduce = 
+        get_op_info("mpi_all_reduce").get_constructor<VarPtr, Var*,NanoString>();
+    return mpi_all_reduce(dout, ns_add);
+}
+
+void MpiAllReduceOp::jit_prepare() {
+    add_jit_define("Tx", x->dtype());
+    add_jit_define("OP", op.to_cstring());
```
[the remark this answers is not in the document]
Comment (cjld) [reply extern/mpi/ops/mpi_all_reduce_op.cc:63]: MPI_MIN,MPI_MAX may added in the future

**Review by cxjyxxme — APPROVED** [review]

**Pull request merged**